Jameswtruher/testgap 004

JamesWTruher opened the pull request.

JamesWTruher committed:
additional tests to close gaps

Move-Item was identical to Rename-Item

JamesWTruher committed:
changing Package tests tags to Feature, as they go off box and should not be CI

JamesWTruher committed:
Add Get-Runspace tests

JamesWTruher committed:
Start adding runspace debugging API tests

JamesWTruher committed:
Rewrite to use $TESTDRIVE which does auto cleanup

JamesWTruher committed:
adding convert-path tests

JamesWTruher committed:
fix up Get-Runspace test to be sure it has a clean environment before starting

JamesWTruher committed:
don't use git to determine where the common directory is

it's not portable for our lab environment, where git may not exist

msftclas commented:
Hi __@JamesWTruher__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Jim Truher). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

lzybkr commented on test/powershell/Modules/Microsoft.PowerShell.Utility/RunspaceCmdlets.Tests.ps1:
```diff
@@ -0,0 +1,39 @@
+        # Clear out all unexpected runspaces
+        # id 1 is the one we started with
+        Get-Runspace |?{$_.id -ne 1} | %{$_.dispose()}
+        $CurrentRunspace = $ExecutionContext.Host.Runspace
```
This seems unwise - e.g. what if Pester used other runspaces, or we started running tests in parallel.

JamesWTruher replied:
tests should clean up after themselves and not pollute, and this code forces that clean up in a rather aggressive manner, however, it may be that it is too aggressive.

---
In reply to: [78446552](https://github.com/PowerShell/PowerShell/pull/2239#discussion_r78446552) [](ancestors = 78446552)

lzybkr commented on test/powershell/Modules/Microsoft.PowerShell.Utility/RunspaceCmdlets.Tests.ps1:
```diff
@@ -0,0 +1,39 @@
+        }
+        It "Get-Runspace should return all runspaces" {
+            $expectedCount = $runspaceCount + 2
+            (get-runspace).Count | should be $expectedCount
```
This test could be fragile if we supported running tests in parallel.

Maybe it's sufficient to test that `Get-Runspace` returns both of the new runspaces.

JamesWTruher replied:
there's no contemplation of running tests in parallel _in the same process_, which is where there would be a problem (and probably with a very large number of our other tests). Running tests in parallel processes won't affect this test at all, however the point is taken

---
In reply to: [78447130](https://github.com/PowerShell/PowerShell/pull/2239#discussion_r78447130) [](ancestors = 78447130)

lzybkr commented on test/powershell/SDK/PSDebugging.Tests.ps1:
```diff
@@ -184,5 +185,94 @@ Describe "PowerShell Command Debugging" -tags "CI" {
+            $ps.dispose()
+        }
+        
+        It "PSStandaloneMonitorRunspaceInfo should throw when called with a null constructor" {
```
>null constructor [](start = 76, length = 16)

I'm not sure what you mean here - sometimes people incorrectly refer to a default constructor as the null constructor, but that's not what you're testing.

JamesWTruher replied:
sure - I can make the language clearer

---
In reply to: [78447508](https://github.com/PowerShell/PowerShell/pull/2239#discussion_r78447508) [](ancestors = 78447508)

JamesWTruher committed:
make the language more clear for the PSStandaloneMonitorRunspaceInfo test

JamesWTruher committed:
removed aggressive runspace cleanup code

also changed get-runspace test when multiple runspaces are present
to look for explicit runspace ids

lzybkr commented:
LGTM

PaulHigin commented on test/powershell/Modules/Microsoft.PowerShell.Utility/RunspaceCmdlets.Tests.ps1:
```diff
@@ -0,0 +1,38 @@
+        $runspace = get-runspace -instanceid $CurrentRunspace.InstanceId
+        $runspace.InstanceId | Should be $ExpectedInstanceId
+    }
+    It "Get-Runspace with runspace Id should return the correct runspace" {
```
>It "Get-Runspace with runspace Id [](start = 4, length = 33)

It seems like we should also include Get-Runspace -Name parameter test as well.

PaulHigin commented on test/powershell/Modules/Microsoft.PowerShell.Utility/RunspaceCmdlets.Tests.ps1:
```diff
@@ -0,0 +1,38 @@
+            $r2 = [runspacefactory]::CreateRunspace()
+        }
+        AfterAll {
+            $r1.Close()
```
>$r1.Close() [](start = 12, length = 11)

You don't need a separate close call.  Dispose will also close the runspace.

PaulHigin commented on test/powershell/SDK/PSDebugging.Tests.ps1:
```diff
@@ -184,5 +185,94 @@ Describe "PowerShell Command Debugging" -tags "CI" {
+            $runspace.Open()
+            $associationId = [guid]::newguid()
+            $runspaceInfo = [PSStandaloneMonitorRunspaceInfo]::new($runspace)
+        }
```
>  } [](start = 6, length = 3)

I believe there also should be a AfterAll {} block to dispose the runspace.

JamesWTruher replied:
quite right

---
In reply to: [78606708](https://github.com/PowerShell/PowerShell/pull/2239#discussion_r78606708) [](ancestors = 78606708)

JamesWTruher committed:
Add get-runspace -name test, remove runspace.close from AfterAll, dispose alone will suffice

JamesWTruher committed:
add runspace.dispose to clean up created runspace

JamesWTruher committed:
Add a test for get-runspace -name and remove the runspace close from AfterAll block

PaulHigin commented:
LGTM

JamesWTruher commented:
it looks like everyone has signed off - can I get a merge?

JamesWTruher committed:
Add get-runspace -name test
